+act #17576 Support serializer with string manifest

patriknw opened the pull request:
- useful when evolution is needed, e.g. Akka Persistence

patriknw committed:
+act #17576 Support serializer with string manifest

* useful when evolution is needed, e.g. Akka Persistence

* docs, comments, cluster-metrics and cluster-tools serializers

patriknw commented:
@ktoso I need to add some more docs, but would like your opinion on the solution first

patriknw commented:
@agemooij you might be interested in this new `SerializerWithStringManifest`. Feedback is welcome.

ahjohannessen commented:
:+1:

ktoso commented on akka-persistence/src/test/scala/akka/persistence/serialization/SerializerSpec.scala:
```diff
@@ -241,6 +316,31 @@ class MyPayloadSerializer extends Serializer {
+      MyPayload2(data = s"${new String(bytes, "UTF-8")}.", n = 0)
+    case other ⇒
+      throw new Exception(s"unexpected manifest [$other]")
+  }
```
:+1:

ahjohannessen commented:
Any chance this gets into 2.3.x?

patriknw commented:
note to self: DistributedPubSubMessageSerializer also needs same treatment

hepin1989 commented:
how about make the manifest as a long/indexed so it could be a long value?

patriknw commented:
@hepin1989 that doesn't work, because we want to support migration from `Serializer` to `SerializerWithStringManifest` in a compatible way. One character encoding should be compact enough.

hepin1989 commented:
make sense.thanks

ahjohannessen commented:
Also, having the type name / string is useful for akka persistence journals like eventstore where you can index on type automatically. A long would be hard to reason about.

ktoso commented on akka-actor/src/main/scala/akka/serialization/Serialization.scala:
```diff
@@ -105,8 +106,34 @@ class Serialization(val system: ExtendedActorSystem) extends Extension {
   /**
+   * Deserializes the given array of bytes using the specified serializer id,
+   * using the optional type hint to the Serializer and the optional ClassLoader ot load it into.
+   * Returns either the resulting object or an Exception if one was thrown.
```
I don't see an "optional ClassLoader" param

patriknw replied:
ah, copy pasta

ktoso commented on akka-actor/src/main/scala/akka/serialization/Serialization.scala:
```diff
@@ -105,8 +106,34 @@ class Serialization(val system: ExtendedActorSystem) extends Extension {
+      serializer match {
+        case s2: SerializerWithStringManifest ⇒ s2.fromBinary(bytes, manifest).asInstanceOf[T]
+        case s1 if manifest == "" || (manifest eq null) ⇒
+          s1.fromBinary(bytes, None).asInstanceOf[T]
```
AFAICS the `manifest eq null` shouldn't really happen in normal operation, as we `if !hasManifest ""` in other places. It's OK to keep the null case here, just making sure that it's not expected due something I missed

patriknw replied:
right, it is better to add a null check were `serializer.manifest` is used

ktoso commented on akka-persistence/src/main/scala/akka/persistence/serialization/MessageSerializer.scala:
```diff
@@ -30,6 +30,8 @@ class MessageSerializer(val system: ExtendedActorSystem) extends BaseSerializer
   val PersistentImplClass = classOf[PersistentImpl]
   val AtLeastOnceDeliverySnapshotClass = classOf[AtLeastOnceDeliverySnap]
 
+  private lazy val serialization = SerializationExtension(system)
```
why lazy?

patriknw replied:
initializaton order of extension and serializer, I think the SerializationExtension creates this instance

ktoso commented on akka-persistence/src/test/scala/akka/persistence/serialization/SerializerSpec.scala:
```diff
@@ -241,6 +316,31 @@ class MyPayloadSerializer extends Serializer {
+  def manifest(o: AnyRef): String = ManifestV2
+
+  def toBinary(o: AnyRef): Array[Byte] = o match {
+    case MyPayload2(data, n) ⇒ s".$data:$n".getBytes("UTF-8")
```
Can we find some place to put the `"UTF-8"`?

patriknw replied:
since we are > jdk6 now, we can use java.nio.charset.StandardCharsets

ktoso replied:
:heart: JDK8

ktoso commented on akka-persistence/src/test/scala/akka/persistence/serialization/SerializerSpec.scala:
```diff
@@ -148,6 +167,57 @@ class MessageSerializerPersistenceSpec extends AkkaSpec(customSerializers) {
+
+        deserialized.payload should be(MyPayload("OldPayload(A)"))
+      }
+    }
```
great test, +1

ktoso commented:
Looks great, on this layer I think that's all we need.
The more complex upcasts we talked about will be performed in the adapter layer.

When writing docs include a mini guide how to use this in persistence's case (like the tests show already).

patriknw commented:
I think I'll update the ordinary serialization docs, but create a separate ticket for "Illustrate real serializer for Persistence, including evolution of event types"

ktoso commented:
Sounds like a good plan +1

ktoso commented:
refs https://github.com/akka/akka/issues/17576

drewhk commented:
LGTM so far

patriknw commented:
@ktoso I have updated:
* fixed your feedback
* added docs (the comprehensive persistence docs will be handled by #17613)
* converted serializers in cluster-metrics and cluster-tools, the reason I can't convert more (e.g. ClusterMessageSerializer) is that they are public, and can therefore not be changed in a binary compatible way (we might change when we do a breaking release)

patriknw commented:
it's a separate commit, if you only want to look at the changes

ktoso commented:
LGTM

agemooij commented:
Hey guys, sorry for the late reply. In our projects so far we have used a custom serializer that prefixes the serialized bytes with a key-version combo, serialized using simple binary encoding. Our deserializers pick the right code based on the key and the version. This maps very well onto this new manifest String so I'm all for it since it saves us some complexity!

I would have preferred a binary encoding but I totally see the need for making this a String. String parsing will probably not add all that much extra deserialization time.

As soon as this stuff becomes available in 2.3.x I'll work it into our upcoming "akka serialization for persistence with support for explicit and compile-time checked migration code" library, code named "stamina" because that first name was a bit long :smile:

Looking forward to show you guys what we have built at Scala Days Amsterdam.

patriknw commented:
Thanks for taking a look at it. Stamina looks like a great library.

We don't plan to include this in 2.3.x, but in 2.4-M2, and 2.4.0 is not that far away.

See you in Amsterdam.

patriknw commented:
Refs #17576

patriknw commented:
also refs #17434
